lib/static-server.ts

```diff
@@ -0,0 +1,153 @@
+import * as fs from 'fs';
+import * as path from 'path';
+import type * as http from 'http';
+import * as url from 'url';
+
+const mimeTypes: { [key: string]: string } = {
+	'.html': 'text/html',
+	'.htm': 'text/html',
+	'.css': 'text/css',
+	'.js': 'application/javascript',
+	'.json': 'application/json',
+	'.xml': 'application/xml',
+	'.txt': 'text/plain',
+	'.md': 'text/plain',
+
+	'.png': 'image/png',
+	'.jpg': 'image/jpeg',
+	'.jpeg': 'image/jpeg',
+	'.gif': 'image/gif',
+	'.svg': 'image/svg+xml',
+	'.ico': 'image/x-icon',
+	'.bmp': 'image/bmp',
+	'.webp': 'image/webp',
+
+	'.woff': 'font/woff',
+	'.woff2': 'font/woff2',
+	'.ttf': 'font/ttf',
+	'.eot': 'application/vnd.ms-fontobject',
+
+	'.zip': 'application/zip',
+	'.tar': 'application/x-tar',
+	'.gz': 'application/gzip',
+
+	'.mp3': 'audio/mpeg',
+	'.wav': 'audio/wav',
+	'.ogg': 'audio/ogg',
+	'.mp4': 'video/mp4',
+	'.webm': 'video/webm',
+};
+
+function isPathSafe(requestedPath: string, baseDir: string): boolean {
+	const resolvedPath = path.resolve(baseDir, requestedPath);
+	const resolvedBase = path.resolve(baseDir);
+	return resolvedPath.startsWith(resolvedBase + path.sep) || resolvedPath === resolvedBase;
+}
+
+function serveFile(
+	res: http.ServerResponse,
+	filePath: string,
+	statusCode = 200,
+	isHeadRequest = false
+): void {
+	fs.stat(filePath, (err, stats) => {
+		if (err || !stats.isFile()) {
+			const notFoundPath = path.resolve(__dirname, '../static/404.html');
+			fs.stat(notFoundPath, (notFoundErr, notFoundStats) => {
+				if (notFoundErr || !notFoundStats.isFile()) {
+					res.writeHead(404, { 'Content-Type': 'text/html' });
+					if (!isHeadRequest) {
+						res.end('<!DOCTYPE html><html><head><title>404 Not Found</title></head><body><h1>Not Found</h1></body></html>');
+					} else {
+						res.end();
+					}
+					return;
+				}
+
+				const notFoundExt = path.extname(notFoundPath);
+				const notFoundMimeType = mimeTypes[notFoundExt] || 'text/html';
+				res.writeHead(404, {
+					'Content-Type': notFoundMimeType,
+					'Content-Length': notFoundStats.size,
+				});
+
+				if (!isHeadRequest) {
+					fs.createReadStream(notFoundPath).pipe(res);
```

Comment:
Please use ``stream.pipeline()``, and attach the error handler to the fs read stream.